lib/apollo-federation/schema.rb

```diff
@@ -62,14 +74,40 @@ def original_query
         @orig_query_object || find_inherited_value(:original_query)
       end
 
+      def compose_directives
+        @compose_directives || find_inherited_value(:compose_directives, [])
+      end
+
+      def links
+        @links || find_inherited_value(:links, [])
+      end
+
+      def all_links
+        imported_directives = IMPORTED_DIRECTIVES
+        imported_directives += IMPORTED_DIRECTIVES_V2_1 if federation_2_1?
+        default_link = {
+          url: 'https://specs.apollo.dev/federation/v2.3',
+          import: imported_directives,
+        }
+        default_link[:as] = default_link_namespace if default_link_namespace != DEFAULT_LINK_NAMESPACE
+        [default_link, *links]
+      end
+
       def federation_2_prefix
-        federation_namespace = ", as: \"#{link_namespace}\"" if link_namespace != DEFAULT_LINK_NAMESPACE
+        schema = "extend schema\n"
 
-        <<~SCHEMA
-          extend schema
-            @link(url: "https://specs.apollo.dev/federation/v2.3"#{federation_namespace}, import: [#{(IMPORTED_DIRECTIVES.map { |directive| "\"@#{directive}\"" }).join(', ')}])
+        all_links.each do |link|
+          schema += "  @link(url: \"#{link[:url]}\""
+          schema += ", as: \"#{link[:as]}\"" if link[:as]
+          schema += ", import: [#{link[:import].map { |d| "\"@#{d}\"" }.join(', ')}]" if link[:import]
+          schema += ")\n"
+        end
+
+        compose_directives.each do |directive|
+          schema += "  @composeDirective(name: \"@#{directive}\")\n"
+        end
 
-        SCHEMA
+        schema + "\n"
```

Comment:
string concatenation with `+` in Ruby is not really a thing; the `<<~SCHEMA .... SCHEMA` is one way of doing it. Here I could also see something like:

```suggestion
schema = ["extend schema"]

all_links.each do |link|
  schema << <<~SCHEMA
    @link(url: \"#{link[:url]}\""
      #{", as: \"#{link[:as]}\"" if link[:as]}
      #{", import: [#{link[:import].map { |d| "\"@#{d}\"" }.join(', ')}]" if link[:import]}
    )
    SCHEMA
end

compose_directives.each do |directive|
  schema << "  @composeDirective(name: \"@#{directive}\")"
end

schema << ""

schema.join("\n")
```

Reply:
I see, the issue with:
```
          schema << <<~LINK
            @link(url: \"#{link[:url]}\""
              #{", as: \"#{link[:as]}\"" if link[:as]}
              #{", import: [#{link[:import].map { |d| "\"@#{d}\"" }.join(', ')}]" if link[:import]}
            )
            LINK
```
is that it inserts blank lines if `link[:as]` or `link[:import]` are not set and we don't end up with a nice one-liner `@link` . Also we need `schema << ''` twice as we want an extra blank line at the end.

I still used `schema.join("\n")` rather than having `\n` in all strings